include/openthread/nat64.h

```diff
@@ -87,6 +87,108 @@ typedef struct otIp4Cidr
     uint8_t      mLength;
 } otIp4Cidr;
 
+typedef struct otIp4Message otIp4Message;
```

Comment:
Can it reuse otMessage like IPv6 messages?

Reply:
Per discussions in https://github.com/openthread/openthread/pull/7836#discussion_r937118309

We use this to only to make sure the packet has the reserved buffer for ipv6.

Then `otIp4MessageFree` is add to make New / Free symmetric, other APIs can use otCastIp4Message for convenient.

Reply:
I don't see there's implementation difference from IPv6 or upper layer messages(CoAP, UDP). All of them share the same otMessage with different reserved buffer. I'm inclined to keep using otMessage for consistency. Thoughts? @abtink

Reply:
As @erjiaqing already mentioned having `otIp4Message` help ensure we have enough reserved space for translation (avoid needing to copy and re-allocate content).

Reply:
I don't have concerns for adding API for allocating buffer with correct reserved bytes for IPv4 packet, which is very similar to already existing APIs to allocate buffers for CoAP and UDP. I just don't see a strong benefit defining a separate otIp4Message for the purpose.

Reply:
Using a different type enforces that the user will use the proper API to allocate (the users cannot pass a `otMessage` to the API if it require a new type of `otIp4Message`). It protects against incorrect use.
 
We can consider removing it (worst-case if the user allocates incorrectly, it still works fine but we may need to re-allocate to make room for headers) but adding new type should not add much overhead.

Reply:
I think this will be the case for CoAP and UDP as well? If users allocate an IPv6 message, but use it for UDP or CoAP, it will also result in this situation. Given we don't define otCoapMessage or otUdpMessage, I guess we don't need to define otIp4Message either. By providing `otIp6NewMessage()`,`otIp4NewMessage()`, `otUdpNewMessage()`, `otCoapNewMessage()`, users should know they need  to allocate messages with correct APIs.

Reply:
I am ok with it either way. It is fine if we want to keep it consistent with previous definitions.

I agree we have similar situation for CoAP and UDP. It may have been a lost opportunity there. For CoAP we actually internally (in core) use different types `Coap::Message` vs `Message` but we did not follow same model in public API. It may be hard to change these now.  The nice thing is that there is not much overhand: We need to define a new opaque type, add couple of more APIs then compiler can do the check at build-time and protects agains incorrect use.

Reply:
Yes, we should use `otMessage` instead of `otIp4Message` if we want to follow the pattern of Coap::Message.

The build time type check sounds nice, and I'm thinking if we can have similar experience but by checking the arguments during runtime?

I'm not sure if print a message (e.g. messages passed to "otNat64Send" must be allocated by "otNewIp4Message") and then crash makes sense.

Reply:
SGTM.

I suggest not doing any run-time checks. It is not incorrect if caller uses `otMessage` allocated differently (it works but would be inefficient).